dali/operators/reader/loader/video/video_file.h

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_READER_LOADER_VIDEO_VIDEO_FILE_H_
+#define DALI_OPERATORS_READER_LOADER_VIDEO_VIDEO_FILE_H_
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/avutil.h>
+#include <libswscale/swscale.h>
+}
+
+#include "dali/core/common.h"
+
+#include <vector>
+#include <string>
+
+
+namespace dali {
+struct IndexEntry {
+  int last_keyframe_id;
+  int pts;
+  bool is_keyframe;
+  bool is_flush_frame;
+};
+
+struct AvState {
+  AVFormatContext *ctx_ = nullptr;
+  AVCodec *codec_ = nullptr;
+  AVCodecParameters *codec_params_ = nullptr;
+  AVCodecContext *codec_ctx_ = nullptr;
+  AVFrame *frame_ = nullptr;
+  AVPacket *packet_ = nullptr;
+  SwsContext  *sws_ctx_ = nullptr;
+  int stream_id_ = -1;
+
+  ~AvState() {
+    sws_freeContext(sws_ctx_);
```

Comment:
What happen if `InitAvState` breaks in the middle? I think then these function will be executed on nullptr.

Reply:
All of them?

Reply:
I checked all of them, they usually have `if (ptr || *ptr) return;`

Reply:
ok

Reply:
The thing is that some of those `libav` calls will free the resources on error on their own. They will set the pointer to NULL then, so the free functions need to be able to handle this.